internal/api/attestationconfigapi/cli/client/reporter.go

```diff
@@ -0,0 +1,358 @@
+			minimalDate = date
+			continue
+		}
+
+		// If the current minimal version has newer versions than the one we just fetched,
+		// update the minimal version to the older version.
+		if isInputNewerThanOtherVersion(variant, *minimalVersion, obj.getVersion()) {
+			v := obj.getVersion().(T)
+			minimalVersion = &v
+			minimalDate = date
+		}
+	}
+
+	return *minimalVersion, minimalDate, nil
+}
+
+func isInputNewerThanOtherVersion(variant variant.Variant, inputVersion, otherVersion any) bool {
+	var result bool
+	actionForVariant(variant,
+		func() {
+			input := inputVersion.(attestationconfigapi.TDXVersion)
+			other := otherVersion.(attestationconfigapi.TDXVersion)
+			result = isInputNewerThanOtherTDXVersion(input, other)
+		},
+		func() {
+			input := inputVersion.(attestationconfigapi.SEVSNPVersion)
+			other := otherVersion.(attestationconfigapi.SEVSNPVersion)
+			result = isInputNewerThanOtherSEVSNPVersion(input, other)
+		},
+	)
+	return result
+}
+
+type apiVersionObject struct {
+	version string          `json:"-"`
+	variant variant.Variant `json:"-"`
+	cached  bool            `json:"-"`
+	snp     attestationconfigapi.SEVSNPVersion
+	tdx     attestationconfigapi.TDXVersion
+}
+
+func (a apiVersionObject) MarshalJSON() ([]byte, error) {
+	var res []byte
+	var err error
+	actionForVariant(a.variant,
+		func() {
+			res, err = json.Marshal(a.tdx)
+		},
+		func() {
+			res, err = json.Marshal(a.snp)
+		},
+	)
+	return res, err
+}
+
+func (a *apiVersionObject) UnmarshalJSON(data []byte) error {
+	errTDX := json.Unmarshal(data, &a.tdx)
+	errSNP := json.Unmarshal(data, &a.snp)
+	if errTDX == nil || errSNP == nil {
+		return nil
+	}
+	return fmt.Errorf("trying to unmarshal data into both TDX and SNP versions: %w", errors.Join(errTDX, errSNP))
+}
+
+// JSONPath returns the path to the JSON file for the request to the config api.
+// This is the path to the cached version in the S3 bucket.
+func (a apiVersionObject) JSONPath() string {
+	if a.cached {
+		return path.Join(reportVersionDir(a.variant), a.version)
+	}
+	return path.Join(attestationconfigapi.AttestationURLPath, a.variant.String(), a.version)
+}
+
+// ValidateRequest validates the request.
+func (a apiVersionObject) ValidateRequest() error {
+	if !strings.HasSuffix(a.version, ".json") {
+		return fmt.Errorf("version has no .json suffix")
+	}
+	return nil
+}
+
+// Validate is a No-Op at the moment.
+func (a apiVersionObject) Validate() error {
+	return nil
+}
+
+func (a apiVersionObject) getVersion() any {
+	var res any
+	actionForVariant(a.variant,
+		func() {
+			res = a.tdx
+		},
+		func() {
+			res = a.snp
+		},
+	)
+	return res
+}
+
+func (a *apiVersionObject) setVersion(version any) {
+	actionForVariant(a.variant,
+		func() {
+			a.tdx = version.(attestationconfigapi.TDXVersion)
+		},
+		func() {
+			a.snp = version.(attestationconfigapi.SEVSNPVersion)
+		},
+	)
+}
+
+func actionForVariant(
+	attestationVariant variant.Variant,
+	tdxAction func(), snpAction func(),
```

Comment:
Why did you choose this over, say, an interface implemented by structs `tdxAction` and `snpAction`?

Reply:
There are many different cases where we need to execute slightly different code paths depending on whether we are uploading a TDX or an SEV-SNP version.
These code paths do not always work with the same data types, hence adding interfaces for this would require lots of extra code.
The goal here was to reduce code changes that have to be done when a new TDX or SEV-SNP is introduced.
With the current approach, adding a new variant should only require adding it to the switch case inside this function.

Reply:
If a new variant is introduced, you need to change all the call sites of `actionForVariant` and add a new func, don't you? I don't think that's less work than adding a new implementation of an interface, but maybe I'm missing something.

Reply:
Only if we introduce a attestation variant that is not based on TDX or SEV-SNP.
E.g. adding `gcp-tdx` or `aws-tdx` would only require adding it to the type switch in `actionForVariant`.
Adding a completely new variant, e.g. something like `arm-realms`, would require more work of course.